pkg/registry/flowcontrol/flowschema/strategy.go

```diff
@@ -0,0 +1,102 @@
+*/
+
+package flowschema
+
+import (
+	"context"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol/validation"
+)
+
+// flowSchemaStrategy implements verification logic for FlowSchema.
+type flowSchemaStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy is the default logic that applies when creating and updating flow schema objects.
+var Strategy = flowSchemaStrategy{legacyscheme.Scheme, names.SimpleNameGenerator}
+
+// NamespaceScoped returns false because all PriorityClasses are global.
+func (flowSchemaStrategy) NamespaceScoped() bool {
+	return false
+}
+
+// PrepareForCreate clears the status of a flow-schema before creation.
+func (flowSchemaStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+	fl := obj.(*flowcontrol.FlowSchema)
+	fl.Status = flowcontrol.FlowSchemaStatus{}
+	fl.Generation = 1
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (flowSchemaStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	newFlowSchema := obj.(*flowcontrol.FlowSchema)
+	oldFlowSchema := old.(*flowcontrol.FlowSchema)
+
+	// Spec updates bump the generation so that we can distinguish between status updates.
+	if !apiequality.Semantic.DeepEqual(newFlowSchema.Spec, oldFlowSchema.Spec) {
+		newFlowSchema.Generation = oldFlowSchema.Generation + 1
+	}
+	newFlowSchema.Status = oldFlowSchema.Status
+}
+
+// Validate validates a new flow-schema.
+func (flowSchemaStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	return validation.ValidateFlowSchema(obj.(*flowcontrol.FlowSchema))
+}
+
+// Canonicalize normalizes the object after validation.
+func (flowSchemaStrategy) Canonicalize(obj runtime.Object) {
+}
+
+func (flowSchemaStrategy) AllowUnconditionalUpdate() bool {
+	return true
+}
+
+// AllowCreateOnUpdate is false for flow-schemas; this means a POST is needed to create one.
+func (flowSchemaStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// ValidateUpdate is the default update validation for an end user.
+func (flowSchemaStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	return validation.ValidateFlowSchema(obj.(*flowcontrol.FlowSchema))
+}
+
+type flowSchemaStatusStrategy struct {
+	flowSchemaStrategy
+}
+
+// StatusStrategy is the default logic that applies when updating flow-schema objects' status.
+var StatusStrategy = flowSchemaStatusStrategy{Strategy}
+
+func (flowSchemaStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	newFlowSchema := obj.(*flowcontrol.FlowSchema)
+	oldFlowSchema := old.(*flowcontrol.FlowSchema)
+	newFlowSchema.Spec = oldFlowSchema.Spec
```

Comment:
Also assign `ObjectMeta`.  Those shouldn't be updatable via status.  Past us made some mistakes, but fixed them for CRDs.

Reply:
added object meta assignment for flow-schema and priority-level. 

> Past us made some mistakes, but fixed them for CRDs.

requiring developers to reset the object-meta here in the status strategy looks fragile.. isnt the update handler a better place to enforce that?

Reply:
I don't personally see the harm in permitting status to update metadata, specifically attributes/labels/owners seem potentially useful. But, I don't really object to doing it this way, either.

Yes, it's brittle and dangerous to do it here (as is incrementing the generation in the main resource), but kind of necessary to not break existing built-ins.

Reply:
Interesting.  I asked about this as a generality, months ago on the slack channel.  I was told that new code should allow an update to the status subresource to _only_ affect the status, IIRC, but that there is deliberate maintainence of alternate behavior for some older types because they established their behavior before the current recommendations were well established.

Reply:
Yeah, I don't think you got wrong advice previously, I just personally don't see the compelling reason that drives people to think that status updates shouldn't touch metadata.